File: api/app/Services/LivrosService.py
from sqlalchemy import Select, func, or_, select
from sqlalchemy.orm import Session

from app.Models.LivroModel import Livro


def _AplicarFiltros(consulta: Select, categoria: str | None, busca: str | None) -> Select:
    if categoria:
        consulta = consulta.where(Livro.Categoria.ilike(f"%{categoria.strip()}%"))
    if busca:
        termo = f"%{busca.strip()}%"
        consulta = consulta.where(
            or_(Livro.Titulo.ilike(termo), Livro.Autores.ilike(termo))
        )
    return consulta


def GetLivros(
    sessao: Session,
    pagina: int = 1,
    tamanhoPagina: int = 20,
    categoria: str | None = None,
    busca: str | None = None,
) -> tuple[list[Livro], int]:
    filtrada = _AplicarFiltros(select(Livro), categoria, busca)

    total = sessao.scalar(select(func.count()).select_from(filtrada.subquery())) or 0
    if total == 0:
        return [], 0

    itens = sessao.scalars(
        filtrada.order_by(Livro.Titulo, Livro.IdLivro)
        .offset((pagina - 1) * tamanhoPagina)
        .limit(tamanhoPagina)
    ).all()

    return list(itens), total


def GetLivroPorId(sessao: Session, idLivro: int) -> Livro | None:
    return sessao.get(Livro, idLivro)


def GetLivrosPorIds(sessao: Session, ids: list[int]) -> dict[int, Livro]:
    if not ids:
        return {}
    livros = sessao.scalars(select(Livro).where(Livro.IdLivro.in_(ids))).all()
    return {livro.IdLivro: livro for livro in livros}


def GetTotalDeLivros(sessao: Session) -> int:
    return sessao.scalar(select(func.count()).select_from(Livro)) or 0

File: src/Livros.Api/Services/LivrosService.cs
using Livros.Dados;
using Livros.Dominio.Entidades;
using Microsoft.EntityFrameworkCore;

namespace Livros.Api.Services;

public class LivrosService(ContextoLivros contexto)
{
    public async Task<(IReadOnlyList<Livro> Itens, int Total)> GetLivrosAsync(
        int pagina,
        int tamanhoPagina,
        string? categoria,
        string? busca,
        CancellationToken cancelamento)
    {
        var consulta = AplicarFiltros(contexto.Livros.AsNoTracking(), categoria, busca);

        var total = await consulta.CountAsync(cancelamento);
        if (total == 0)
        {
            return ([], 0);
        }

        var itens = await consulta
            .OrderBy(livro => livro.Titulo)
            .ThenBy(livro => livro.IdLivro)
            .Skip((pagina - 1) * tamanhoPagina)
            .Take(tamanhoPagina)
            .ToListAsync(cancelamento);

        return (itens, total);
    }

    public Task<Livro?> GetLivroPorIdAsync(long idLivro, CancellationToken cancelamento) =>
        contexto.Livros.AsNoTracking().FirstOrDefaultAsync(livro => livro.IdLivro == idLivro, cancelamento);

    public async Task<Dictionary<long, Livro>> GetLivrosPorIdsAsync(
        IReadOnlyCollection<long> ids,
        CancellationToken cancelamento)
    {
        if (ids.Count == 0)
        {
            return [];
        }

        return await contexto.Livros
            .AsNoTracking()
            .Where(livro => ids.Contains(livro.IdLivro))
            .ToDictionaryAsync(livro => livro.IdLivro, cancelamento);
    }

    private static IQueryable<Livro> AplicarFiltros(IQueryable<Livro> consulta, string? categoria, string? busca)
    {
        if (!string.IsNullOrWhiteSpace(categoria))
        {
            var padrao = Filtros.MontarPadraoDeBusca(categoria);
            consulta = consulta.Where(livro => EF.Functions.ILike(livro.Categoria!, padrao));
        }

        if (!string.IsNullOrWhiteSpace(busca))
        {
            var padrao = Filtros.MontarPadraoDeBusca(busca);
            consulta = consulta.Where(livro =>
                EF.Functions.ILike(livro.Titulo, padrao) || EF.Functions.ILike(livro.Autores!, padrao));
        }

        return consulta;
    }

}
